pkg/agent/svid/rotator.go

```diff
@@ -124,15 +136,79 @@ func (r *rotator) SetRotationFinishedHook(f func()) {
 }
 
 func (r *rotator) rotateSVIDIfNeeded(ctx context.Context) (err error) {
-	if rotationutil.ShouldRotateX509(r.clk.Now(), r.state.Value().(State).SVID[0]) {
-		err = r.rotateSVID(ctx)
+	state, ok := r.state.Value().(State)
+	if !ok {
+		return fmt.Errorf("unexpected value type: %T", r.state.Value())
 	}
-	if r.rotationFinishedHook != nil {
-		r.rotationFinishedHook()
+
+	if rotationutil.ShouldRotateX509(r.clk.Now(), state.SVID[0]) {
+		if state.Reattestable && fflag.IsSet(fflag.FlagReattestToRenew) {
+			err = r.reattest(ctx)
+		} else {
+			err = r.rotateSVID(ctx)
+		}
+
+		if err == nil && r.rotationFinishedHook != nil {
+			r.rotationFinishedHook()
+		}
 	}
+
 	return err
 }
 
+// reattest goes through the full attestation process with the server and gets a new SVID.
+func (r *rotator) reattest(ctx context.Context) (err error) {
+	counter := telemetry_agent.StartReattestAgentCall(r.c.Metrics)
+	defer counter.Done(&err)
+
+	// Get the mtx before starting the reattestation
+	// In this way, the client do not create new connections until the new SVID is received
+	r.rotMtx.Lock()
+	defer r.rotMtx.Unlock()
+	r.c.Log.Debug("Reattesting node")
+
+	bundle, err := r.getBundle()
+	if err != nil {
+		return err
+	}
+
+	key, err := r.generateKey(ctx)
+	if err != nil {
+		return err
+	}
+
+	csr, err := util.MakeCSRWithoutURISAN(key)
+	if err != nil {
+		return err
+	}
+
+	conn, err := r.serverConn(ctx, bundle)
```

Comment:
We should close this connection after use (e.g. via defer, like the renew code).